content/blog/2020-11-29-tracemonkey/index.md

```diff
@@ -0,0 +1,402 @@
+described above, as well as the inability to trace across recursive
+function calls, into regular expression subtitutions with lambda
+functions, or through the usage of the `eval` function.
+
+The following graph shows the fraction of time that the benchmarks
+spent in each state from the high-level state machine:
+
+![breakdown of time spent in each state](results_breakdown.png)
+
+As can be seen, the benchmarks that performed the best spent almost
+all of their time running compiled traces, while those that performed
+the worse either were unable to compile any traces at all (due to
+languages features where tracing was not yet supported) or got bogged
+down tracing/compiling (possibly due to the issue with trace
+stitching). Thus the authors suspect that the performance will be
+superior and perhaps exceed that of V8 or SFX once these issues are
+ironed out. This paper is largely a proof of concept for tracing JITs
+with type specialization, and it is not necessarily expected to
+produce an implementation without quirks.
+
+## Discussion
+
+TraceMonkey was
+[merged](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/Releases/1.8)
+into SpiderMonkey in version 1.8.1 and
+[released](https://wiki.mozilla.org/JavaScript:TraceMonkey) as part of
+Firefox 3.5. Clearly the work was deemed successful and, eventually,
+production-ready.
+
+That being said, TraceMonkey was
+[replaced](https://wiki.mozilla.org/JaegerMonkey) by a different
+implementation, JaegerMonkey, in 2010. JaegerMonkey is a method-based
+JIT with type specialization, rather than a tracing JIT. The current
+implementation is [IonMonkey](https://wiki.mozilla.org/IonMonkey),
+which is also a method-based JIT with type specialization. It seems
+that the tracing never caught on, but the type specialization did.
+
+The two main contributions of this paper are:
+
+1. how to create a tracing JIT for a dynamic language like JavaScript
+
+2. how to integrate type specialization in a JIT for a dynamic
+   language like JavaScript
+
+While both are very valuable, it seems that the type specialization is
+the more durable contribution.
+
+### Discussion questions
+
+ - TraceMonkey has one pass of lightweight optimizations and a trivial
+   register allocator, whereas more modern engines like V8 include
+   several passes of increasingly complex optimizations. What kind of
+   optimizations should be performed by JIT compilers for dynamic
+   languages? When are the tradeoffs between compilation time and
+   improved performance worthwhile?
+
+ - Modern JavaScript engines use parallelization to improve
+   performance, much like the authors of this paper described doing
+   trace stitching in parallel. Yet TraceMonkey does not rely on
+   parallelization and other JITs, like the JVM (at least in the
+   past), don't seem to either. Are modern JavaScript JIT
+   implementations tractable in non-parallel environments?
```

Comment:
To clarify, are you talking about running the _compiler_ in parallel or generating _compiled code_ that runs in parallel?